numpy/core/tests/test_regression.py

```diff
@@ -2214,6 +2214,20 @@ def test_reshape_size_overflow(self):
             new_shape = (2, 7, 7, 43826197)
         assert_raises(ValueError, a.reshape, new_shape)
 
+    def test_invalid_structured_dtypes(self):
+        # gh-2865
+        # the following invalid dtypes were causing segfaults
+        assert_raises(ValueError, np.dtype, ('O', [('name', 'i8')]))
+        assert_raises(ValueError, np.dtype, ('i8', [('name', 'O')]))
+        # these were causing SystemErrors
+        assert_raises(ValueError, np.dtype,
+                      ('i', {'name': ('i', 0, 'title', 'oops')}))
+        assert_raises(ValueError, np.dtype,
+                      ('i', {'name': ('i', 'wrongtype', 'title')}))
+        # this one is allowed as a special case though
+        a = np.ones(1, dtype=('O', [('name', 'O')]))
```

Comment:
You're right, #2798 suggests that some people (eg h5py) have wanted to be able to create `dtype(("O", [("name", "O")]))`, in order to add a field name. I would have liked to simply disallow that, but maybe we can't. 

Its probably safe to assume no-one is doing anything like `dtype(('O,O','O,i8'))`, so my suggestion is to disallow all cases involving objects (by RECHCK on both elements of the tuple) except the special case you already have for a single field. 

I am a little tempted to add a deprecation warning for a release cycle, but it's probably such a rare problem that I would rather just go ahead with the change. But since it can break code, we should clearly document it. So please add something to `release/1.13.0-notes.rst` decsribing what will break.

By the way, the example causing problems in h5py is in https://github.com/h5py/h5py/issues/217. I tried the example, but it looks like it no longer involves the special case dtype. Also, incidentally, after #6053 gets in they will be able to write `np.zeros(3, dtype='O').astype([('name', 'O')])` instead (this is currently disallowed). So there will be a better syntax for adding a field.

Also, by the way, I once wrote a lot of code in #5548 (in `_check_field_overlap`) to carefully check whether views like `'O,O'` to `'O,i8'` would clobber pointers or not, by searching out the positions of all the pointers. But it was a huge complicated affair we got rid of later because it was too slow. That's why I favor the stupid solution of simply disallowing views involving objects.